### PR TITLE
HW04 by Kate Fedotova

### PR DESCRIPTION
**Pull request by fedotovakate02**

### CODE REVIEW

File: HW04/Kate Fedotova/arr.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    (0..length - 1).each do |item_idx|
+      yield(self[item_idx])
+    end
+
+    self
+  end
+
+  def my_map
+    return clone.to_enum unless block_given?
```
Comment (vladhilko) [inline HW04/Kate Fedotova/arr.rb:13]: Why do you need `clone` here? Where did you find it? 🤔 😅
```suggestion
    return to_enum unless block_given?
```

File: HW04/Kate Fedotova/arr.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class Array
+  def my_each
+    return self unless block_given?
+
+    (0..length - 1).each do |item_idx|
```
Comment (vladhilko) [inline HW04/Kate Fedotova/arr.rb:5]: What is `item_idx`? 🤔 Any other ideas for this name? 😬

**Comment by vladhilko** [comment]
Please, run `rubocop-rspec` 😔

File: HW04/Kate Fedotova/arr.rb (hunk 1 above)
Comment (fedotovakate02) [reply HW04/Kate Fedotova/arr.rb:13]: I found it in internet. Colne creates a copy of the array.

File: HW04/Kate Fedotova/arr.rb (hunk 2 above)
Comment (fedotovakate02) [reply HW04/Kate Fedotova/arr.rb:5]: I invent other name

File: HW04/Kate Fedotova/array_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+require 'rspec'
+require 'rubocop-rspec'
+require_relative 'arr.rb'
+
+# rubocop:disable Metrics/BlockLength
+describe Array do
+  describe 'my_each' do
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:7]: We use this style to describe method name http://www.betterspecs.org/#describe
```suggestion
  describe '#my_each' do
```

File: HW04/Kate Fedotova/array_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    let(:arr) { [1, 2, 3, 4] }
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:8]: You can use subject here, it's the same but will be more correct
```suggestion
    subject(:arr) { [1, 2, 3, 4] }
```

File: HW04/Kate Fedotova/array_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    it 'has block' do
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:10]: Can you add a context? Something like this http://www.betterspecs.org/#contexts
```suggestion
    context 'when has block' do
      it 'returns correct result' do
```

File: HW04/Kate Fedotova/array_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      new_arr = []
+      arr.my_select { |el| new_arr << el }
+      expect(new_arr).to eq([1, 2, 3, 4])
+    end
+  end
+
+  describe 'my_map' do
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:17]: The same comments for this test

File: HW04/Kate Fedotova/array_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    it 'has block' do
+      expect(arr.my_map { |el| el + 1 }).to eq([2, 3, 4, 5])
+    end
+
+    it 'without block' do
+      expect(arr.my_select).to be_an_instance_of(Enumerator)
+    end
+  end
+
+  describe 'my_select' do
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:29]: The same comments for this test

File: HW04/Kate Fedotova/array_spec.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
+require 'rspec'
+require 'rubocop-rspec'
+require_relative 'arr.rb'
+# rubocop:disable Metrics/BlockLength
+describe Array do
+  describe '#my_each' do
+    subject(:arr) { [1, 2, 3, 4] }
+
+    context 'when has block' do
+      it 'return arr items +1'
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:10]: does it work? 🤔

File: HW04/Kate Fedotova/array_spec.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      new_arr = []
+      arr.my_select { |el| new_arr << el }
+      expect(new_arr).to eq([1, 2, 3, 4])
```
Comment (vladhilko) [inline HW04/Kate Fedotova/array_spec.rb:13]: Shouldn't we use this way? 🤔 
```suggestion
      it 'return arr items +1' do 
        new_arr = []
        arr.my_select { |el| new_arr << el }
        expect(new_arr).to eq([1, 2, 3, 4])
      end
```

File: HW04/Kate Fedotova/array_spec.rb (hunk 8 above)
Comment (fedotovakate02) [reply HW04/Kate Fedotova/array_spec.rb:10]: yes, it works

**Pull request merged**